### PR TITLE
atmel-samd: Introduce a nvm module for non-volatile byte-level memory access.

### LINKED ISSUE #160: 256b scratch space thats writeable from circuitpython
raw bytes rather than filesystem

### PR DESCRIPTION
**Pull request by tannewt**
This allows for persisting small configuration values even when the file system
is read-only from CircuitPython.

Fixes #160

### CODE REVIEW

File: shared-bindings/nvm/ByteArray.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+STATIC mp_obj_t nvm_bytearray_make_new(const mp_obj_type_t *type,
+        mp_uint_t n_args, mp_uint_t n_kw, const mp_obj_t *args) {
+    return mp_const_none;
+}
+
+//|   .. method:: __len__()
+//|
+//|     Return the length. This is used by (`len`)
+//|
+STATIC mp_obj_t nvm_bytearray___len__(mp_obj_t self_in) {
```
Comment (dhalbert) [inline shared-bindings/nvm/ByteArray.c]: This should work, but `__len__()` doesn't seem to work right in Circuit/MicroPython:
```
>>> len(microcontroller.nvm)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'ByteArray' has no len()

>>> len(bytearray(7))
7

>>> microcontroller.nvm.__len__()
256

>>> bytearray(7).__len__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'bytearray' object has no attribute '__len__'
>>> [1,2].__len__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute '__len__'

```
It doesn't work in the `unix` build either, so it's not some idiosyncrasy of `atmel-samd`

**Comment by dhalbert** [comment]
When I tested this on a CPX, I saw `CIRCUITPYTHON1` and the crystal-less usb calibration value in the first bytes of `microcontroller.nvm`. Do you think this was left over from a previous non-`nvm` firmware load, or is `main.c` overwriting nvm with the cal data? I'm not sure of the order of firmware, nvm and calibration areas in the flash.

File: atmel-samd/main.c [hunk 2]
```diff
@@ -568,6 +568,27 @@ safe_mode_t samd21_init(void) {
...
+        for (uint8_t row = 0; row < TOTAL_INTERNAL_FLASH_SIZE / NVMCTRL_ROW_SIZE; row++) {
+            uint32_t new_row_address = INTERNAL_FLASH_MEM_SEG1_START_ADDR + row * NVMCTRL_ROW_SIZE;
+            nvm_erase_row(new_row_address);
+            nvm_write_buffer(new_row_address,
+                             (uint8_t*) (new_row_address + CIRCUITPY_INTERNAL_EEPROM_SIZE),
+                             NVMCTRL_ROW_SIZE);
+        }
+        #endif
+        // Mark the last section as eeprom now.
+        fuses.eeprom_size = NVM_EEPROM_EMULATOR_SIZE_256;
```
[the remark this answers is not in the document]
Comment (tannewt) [reply atmel-samd/main.c]: Added a computation of the enum value so it is flexible.

File: shared-bindings/nvm/ByteArray.c [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/nvm/ByteArray.h"
+
+//| .. currentmodule:: nvm
+//|
+//| :class:`ByteArray` -- Presents a stretch of non-volatile memory as a bytearray.
+//| ================================================================================
+//|
+//| Non-volatile memory is avialble as a byte array that persists over reloads
```
[the remark this answers is not in the document]
Comment (tannewt) [reply shared-bindings/nvm/ByteArray.c]: Done.

File: shared-bindings/nvm/ByteArray.c (hunk 1 above)
Comment (tannewt) [reply shared-bindings/nvm/ByteArray.c]: Fixed. Its unary_op internally :-/

**Comment by tannewt** [comment]
Yup, I moved the location of the config data to before the eeprom section. It needs to be at the end of memory for the nvm controller to understand where it is.

**Pull request merged**